accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class RemovedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class ObsoletedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+        public string Url { get; set; }
+    }
+}
+```
+
+### Platform context
+
+To determine the platform context of the call site the analyzer must consider
+the following two configurations:
+
+1. **Call-site**. Application of `AddedInOSPlatformAttribute` to the containing
+   member, type, module, or assembly.
+
+2. **Project** The project TFMs and `TargetPlatformMinVersion`. If the project
+   has an platform-specific TFM (such as `net5.0-ios13.0`) we consider it to be
+   on the specified platform. If `TargetPlatformMinVersion` is set, we assume
+   that version, otherwise we assume the version as specified by the TFM.
+
+If (2) is platform neutral (e.g. `net5.0`), the platform context is entirely
+determined by (1).
+
+If (2) is platform-specific it is combined with (1):
+
+* If the platforms between and (2) are disjoint (such as iOS and Windows), the
+  information from (1) is discarded.
+* Otherwise the platform context's minimum version is the maximum of the
+  versions determined by (1) and (2).
+
+### Diagnostics
+
+The analyzer can produce three diagnostics:
+
+1. '{API}' requires {OS} {OS version} or later.
+2. '{API}' has been deprecated since {OS} {OS version}.
+3. '{API}' has been removed since {OS} {OS version}.
+
+### Automatic suppression via platform guards
+
+Diagnostic (1) can be suppressed via a call to
+`RuntimeInformation.IsOSPlatformOrLater()`. This should work for for simple cases
```

Comment:
I assume we only intend to do intra-method and not inter-method analysis?  What's the intended gesture on the part of the developer if they have code like this:
```C#
public void DoWork()
{
    if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) DoWork_Windows();
    else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) DoWork_Linux();
    else { ... }
}

private void DoWork_Windows()
{
    ... // call Windows-specific API
}

private void DoWork_Linux()
{
    ... // call Linux-specific API
}
```

Reply:
Yes, no intra-method analysis. I think attributes is the best way to deal with this.